use commerce api v2.

jimabramson opened the pull request:
XCOM-213 XCOM-214
@rlucioni  @wedaly  please review and enjoy.

still calling this 'WIP' because i expect a few test failures and a decent amount of corrections.

wedaly commented on lms/djangoapps/commerce/urls.py:
```diff
@@ -4,10 +4,10 @@
 urlpatterns = patterns(
     '',
-    url(r'^orders/$', OrdersView.as_view(), name="orders"),
+    url(r'^baskets/$', BasketsView.as_view(), name="baskets"),
```
Is there a risk that clients might still try to contact the old end-point if they're using cached JS of the logistration page?  It might make sense to leave the old end-point for an initial release, then remove it in the following release.

wedaly replied:
Actually, looking more closely at that code, I'm guessing you'll need to update this as well:
https://github.com/edx/edx-platform/blob/master/lms/static/js/student_account/enrollment.js#L12

jimabramson replied:
@wedaly thanks, both good calls.
I will look at making an adapter to keep orders/ working with the v2 api.
The bok choy failures actually traced back to that same line in enrollment.js, I've fixed that.

wedaly commented on lms/djangoapps/commerce/views.py:
```diff
@@ -103,28 +103,27 @@ def post(self, request, *args, **kwargs):  # pylint: disable=unused-argument
-
-                msg = Messages.ORDER_INCOMPLETE_ENROLLED.format(order_number=order_number)
-                return DetailResponse(msg, status=HTTP_202_ACCEPTED)
+                # TODO how does the user recover from this?
```
On the logistration page, which auto-enrolls the user in a free track, we don't display any errors.  The assumption is that it's important to successfully authenticate the user and get them to a track selection page where they might continue to payment, and we don't want to block that process.

On the payment page, I believe we show an error message (using ErrorModel and ErrorView defined using BackBone views and models).  There probably isn't a whole lot the user can do in this case except to contact support.

jimabramson replied:
OK, it sounds like the js client knows what to do with the error; I'm removing the TODO.

wedaly commented on lms/djangoapps/verify_student/views.py:
```diff
@@ -724,7 +731,7 @@ def create_order(request):
 
     if current_mode.sku:
-        return create_order_with_ecommerce_service(request.user, course_id, current_mode)
+        return create_order_with_ecommerce_service(request.user, course_id, current_mode, request.POST.get('processor'))
```
What happens if the JS client doesn't include a payment processor in the request?

jimabramson replied:
@wedaly this should only happen for the stale cached js right?  I'm changing it to default to `settings.CC_PROCESSOR_NAME` as a fallback.

rlucioni replied:
@jimabramson given that you're using `get()` here, that shouldn't be necessary. `create_order_with_ecommerce_service()` uses the API client's `create_basket()` under the hood, to which `payment_processor` is an optional argument defaulting to `None`. On the Otto side, if a payment processor name is required but none is provided, the default payment processor (specified in configuration) is used.

wedaly commented on lms/djangoapps/verify_student/views.py:
```diff
@@ -724,7 +731,7 @@ def create_order(request):
         return HttpResponseBadRequest(_("No selected price or selected price is below minimum."))
 
     if current_mode.sku:
```
Just a side note on lines 675-702 above: it looks like the photo verification code is left over from before we launched the decoupled verified payment flow!  Definitely not required for this PR, but at some point it would be great to clean this up :)

jimabramson replied:
@wedaly created XCOM-257 to track this.

rlucioni commented on lms/djangoapps/commerce/views.py:
```diff
@@ -103,28 +103,26 @@ def post(self, request, *args, **kwargs):  # pylint: disable=unused-argument
+                honor_mode.sku,
+                payment_processor="cybersource",
+            )
+            if response_data['payment_data'] is not None:
```
I'd assign the result of `response_data['payment_data']` to a variable, say `payment_data`, so you don't need to retrieve it twice. It'd also be a little more Pythonic if this was:
```python
if payment_data:
```

jimabramson replied:
changed.

rlucioni commented on lms/djangoapps/commerce/views.py:
```diff
@@ -103,28 +103,26 @@ def post(self, request, *args, **kwargs):  # pylint: disable=unused-argument
             else:
-                # TODO Before this functionality is fully rolled-out, this branch should be updated to NOT enroll the
-                # user. Enrollments must be initiated by the E-Commerce API only.
-                self._enroll(course_key, user)
```
Are we ready to get rid of this failsafe?

jimabramson replied:
Since you raise this, after some consideration, I'm going to put it back.  I was originally thinking this failsafe wasn't going to be a good idea when the mode being handled is a paid mode.  I've since become clear on the fact that this endpoint doesn't have to handle paid modes yet.

rlucioni commented on lms/djangoapps/commerce/api.py:
```diff
@@ -60,26 +60,52 @@ def get():
+
+        Note that while the API supports deferring checkout to a separate step,
+        this client does not currently need that capability, so that case is
+        not supported.
```
It may also be worth noting here that in addition to deferred checkout, the API supports adding more than one product at a time to the user's basket.

jimabramson replied:
agree, added.

rlucioni commented on lms/djangoapps/shoppingcart/tests/test_views.py:
```diff
@@ -918,7 +918,12 @@ class mock_get_order(object):  # pylint: disable=invalid-name
+            'status': OrderStatus.COMPLETE,
+            'number': EcommerceApiTestMixin.ORDER_NUMBER,
+            'payment_processor': 'cybersource',
+            'payment_parameters': {'orderNumber': EcommerceApiTestMixin.ORDER_NUMBER},
```
This test data doesn't appear to be representative of the data returned from the ecommerce API; payment information is keyed under `payment_data`, separate from the information keyed under `order`.

rlucioni replied:
I'm sorry, I see that this is meant to mock a serialized order. Assuming this is the case, `payment_processor` and `payment_parameters` are no longer part of serialized orders, so you should be able to remove them.

jimabramson replied:
removed.

rlucioni commented on lms/djangoapps/verify_student/tests/test_views.py:
```diff
@@ -871,6 +873,14 @@ def _post(self, data):
         url = reverse('verify_student_create_order')
```
This test class is named `TestCreateOrder` and the tests it contains are described as "Tests for the create order view."  Should this be updated? Is the URL being POSTed to here still correct?

jimabramson replied:
we previously discussed leaving verify_student's /order/ endpoint named as it is, but we can revisit that decision that if this doesn't feel right.

rlucioni replied:
That's fine, I remember our conversation now. To preserve our sanity, it might be worth adding a line to the docstring for this test class explaining that verify_student's `create_order` sometimes uses the old shoppingcart but will instead call out to Otto if the requested course mode is associated with a SKU.

rlucioni commented on lms/djangoapps/verify_student/tests/test_views.py:
```diff
@@ -871,6 +873,14 @@ def _post(self, data):
+        """ DRY helper. """
+        data = json.loads(response.content)
+        self.assertEqual(set(data.keys()), PAYMENT_DATA_KEYS)
+        payment_form_data = data['payment_form_data']
```
Should this be `data['payment_data']['payment_form_data']`?

jimabramson replied:
as discussed offline, the 'payment_data' dictionary is the whole of the response from this endpoint, so this doesn't need to be changed.

rlucioni commented on lms/djangoapps/verify_student/tests/test_views.py:
```diff
@@ -885,12 +895,14 @@ def test_create_order_already_verified(self):
+            {
+                'override_custom_receipt_page': "http://testserver/shoppingcart/postpay_callback/",
+                'merchant_defined_data1': unicode(self.course.id),
+                'merchant_defined_data2': "verified",
```
It may not be important here, but these "merchant-defined data" fields aren't currently sent back by the ecommerce API.

jimabramson replied:
@rlucioni yeah - this response assertion is coupled both to the legacy shoppingcart and to the cybersource processor.  I think what's really being tested here is that it doesn't try to trigger the verification process, so maybe I should just remove the detailed response data assertions entirely.  Thoughts?

rlucioni replied:
I agree with paring down the test such that it verifies the expected behavior.

That said, looking at this test and others like it more closely, I don't think there should be a difference in behavior at this point if the user is verified. Formerly, users couldn't make it to payment unless they were already verified. Currently, this part of the flow should behave the same for verified and non-verified users alike.

jimabramson replied:
Good point.  I ended up substantially rewriting this test so that it would run the same requests against both backends.  I did remove everything that differentiates between verified and non-verified users.  The code/tests in this package need even more cleanup work but that's best saved for subsequent PRs.

rlucioni commented on lms/djangoapps/verify_student/tests/test_views.py:
```diff
@@ -1099,7 +1124,7 @@ def test_create_order_success(self):
         )
         json_response = json.loads(response.content)
-        self.assertIsNotNone(json_response.get('orderNumber'))
+        self.assertIsNotNone(json_response['payment_form_data'].get('orderNumber'))  # TODO not canonical
```
I might be confused about the behavior being tested here. Is this meant to be data returned by the ecommerce API? The payment form data sent back by the ecommerce API doesn't include an `orderNumber` key.

jimabramson replied:
this test runs against the shoppingcart backend and assumes the cybersource processor, which has 'orderNumber' in its expected form data params.  I've left it alone except to correct the location in the response payload where the key is expected to be found.

rlucioni replied:
Thanks for the context.

rlucioni commented on lms/djangoapps/verify_student/views.py:
```diff
@@ -644,15 +652,14 @@ def _check_enrollment(self, user, course_key):
+        response_data = api.create_basket(user, course_mode.sku, processor)  # TODO get correct processor from request
         # Pass the payment parameters directly from the API response.
-        return HttpResponse(json.dumps(data['payment_parameters']), content_type='application/json')
+        return HttpResponse(json.dumps(response_data['payment_data'] or {}), content_type='application/json')
```
Assuming this is meant to prevent exceptions from being raised if `payment_data` is missing from the response, this would be safer as `response_data.get('payment_data') or {}`.

jimabramson replied:
actually I assume (based on [this documentation](https://github.com/edx/ecommerce/blob/master/ecommerce/extensions/api/v2/views.py#L71-L118) that `payment_data` will always be present in the response, but it may be None.   This is still a bit of a punt however, I think it might be better to change to `payment_data or order_data`.

rlucioni replied:
I'm not sure returning `order` if `payment_data` is `None` is any better than returning an empty dict. Callers of this function expect payment data to be returned, so returning anything other than payment data, whether that's order data or an empty dict, will cause a failure later down the line, when the user tries to reach CyberSource. The following might even be fine:
```python
return HttpResponse(json.dumps(response_data.get('payment_data')), content_type='application/json')
```
If payment data is missing, this will cause a failure in [the JS responsible for constructing the form](https://github.com/edx/edx-platform/pull/7777/files#diff-fd4675021d17a3af130c8ded0c74a05cR126) POSTed to CyberSource.

What do you think? Ideally, we'd raise an exception and handle it gracefully.

jimabramson replied:
OK, agreed, since payment data is the only response ever expected here, I will make the suggested change.

rlucioni commented on lms/djangoapps/verify_student/views.py:
```diff
@@ -724,7 +731,9 @@ def create_order(request):
     if current_mode.sku:
-        return create_order_with_ecommerce_service(request.user, course_id, current_mode)
+        return create_order_with_ecommerce_service(
+            request.user, course_id, current_mode, request.POST.get('processor', settings.CC_PROCESSOR_NAME)
```
I don't think it's necessary to default to `settings.CC_PROCESSOR_NAME` in the event that no processor name is included in the request. `create_order_with_ecommerce_service()` uses the API client's `create_basket()` under the hood, to which `payment_processor` is an optional argument defaulting to `None`. On the Otto side, if a payment processor name is required but none is provided, the default payment processor (specified in configuration) is used.

jimabramson replied:
makes sense, changed.

rlucioni commented on lms/static/js/verify_student/views/make_payment_step_view.js:
```diff
@@ -98,7 +109,7 @@ var edx = edx || {};
 
-        handleCreateOrderResponse: function( paymentParams ) {
+        handleCreateOrderResponse: function( paymentData ) {
             // At this point, the order has been created on the server,
```
This is not longer always the case. If the ecommerce service is being used, a basket has been created and frozen on the server, but no order has been created.

jimabramson replied:
good catch, fixed.

rlucioni commented on lms/static/js/verify_student/views/make_payment_step_view.js:
```diff
@@ -59,23 +60,33 @@ var edx = edx || {};
         },
 
-        createOrder: function() {
+        createOrder: function(event) {
```
Seeing as how this is the "top of the chain," so to speak, this might be a good place to note that a basket will be created instead of an order if Otto is in use.

jimabramson replied:
makes sense, done.

rlucioni commented on lms/djangoapps/verify_student/views.py:
```diff
@@ -644,15 +652,14 @@ def _check_enrollment(self, user, course_key):
         # Make an API call to create the order and retrieve the results
-        _order_number, _order_status, data = api.create_order(user, course_mode.sku)
-
+        response_data = api.create_basket(user, course_mode.sku, processor)  # TODO get correct processor from request
```
I think you've addressed the TODO here.

jimabramson commented:
updated:
* verify_student.tests.test_views.TestCreateOrder has been rewritten to run the same tests against both Otto & legacy shoppingcart
* added server-side handling to ensure requests for /commerce/orders/ and /verify_student/create_order/ continue to work with outdated js clients during update
* adding preliminary handling for dynamically generated button text, as discussed offline
* improved some documentation
* addressing remaining inline comments from @rlucioni

I expect to make one more update to this PR today with changes to the template/UX only.

jimabramson commented:
test this please.   preferably with less totally unrelated failures.

rlucioni commented on lms/djangoapps/commerce/api.py:
```diff
@@ -60,26 +60,52 @@ def get():
-            sku     --  SKU of the course seat being ordered.
+            user                     --  User for which the basket should be created.
+            sku                      --  SKU of the course seat being ordered.
+            payment_processor        --  Name of the payment processor to use.
```
Minor: I'm not sure where this docstring format came from (I know it was here before you started working on this), but it's not consistent with either the Google-style docstrings we use elsewhere, or the DRF API format Andy's sent an email around about. I'd prefer it if the docstring was consistent with Google-style.

rlucioni commented on lms/djangoapps/commerce/urls.py:
```diff
@@ -4,10 +4,12 @@
-    url(r'^orders/$', OrdersView.as_view(), name="orders"),
+    url(r'^baskets/$', BasketsView.as_view(), name="baskets"),
     url(r'^checkout/cancel/$', checkout_cancel, name="checkout_cancel"),
+    # For backwards compatibility with js clients during intial release
```
Including a unique identifier (e.g., XCOM-214) in a comment near this and other changes introduced for backwards compatibility would make the clean-up process easier.

rlucioni commented on lms/djangoapps/verify_student/views.py:
```diff
@@ -724,35 +765,23 @@ def create_order(request):
+            request.POST.get('processor')
+        )
+    else:
+        payment_data = checkout_with_shoppingcart(request, request.user, course_id, current_mode, amount)
```
Nice, I like how clear this is.

rlucioni commented on lms/djangoapps/verify_student/views.py:
```diff
@@ -724,35 +765,23 @@ def create_order(request):
+        payment_data = checkout_with_shoppingcart(request, request.user, course_id, current_mode, amount)
+
+    if 'processor' not in request.POST:
+        # To be removed after release.
```
This would be another good place to mark with a unique identifier (e.g., XCOM-214).

rlucioni commented:
I've made a few additional, albeit minor, comments. Those aside, :+1: . Thanks for your hard work on this!

wedaly commented on lms/djangoapps/commerce/views.py:
```diff
@@ -103,28 +103,31 @@ def post(self, request, *args, **kwargs):  # pylint: disable=unused-argument
-                return DetailResponse(msg, status=HTTP_202_ACCEPTED)
+                msg = u'Unexpected response from basket endpoint.'
+                log.error(
+                    msg + u' Could not enroll user %(username)s in course %(course_id)s.',
```
Would it be safer to use `user.id` for privacy reasons?

jimabramson replied:
i'll change it for now, though i'd like to better understand what the privacy concern is (we can discuss offline).

wedaly commented on lms/templates/verify_student/make_payment_step.underscore:
```diff
@@ -98,11 +98,10 @@
-      <%- gettext( "Continue to payment" ) %> ($<%- minPrice %>)
-    </a>
+    <span>price: $<%- minPrice %></span>
+    <!-- payment buttons will go here -->
```
Pretty sure this comment will show up in the output HTML.  Would it be possible to use JS comments instead?
```
<%
// payment buttons will go here
%>
```

jimabramson replied:
good call, fixing.

jimabramson replied:
wait, does underscore strip that out when it renders?

wedaly commented:
Small comments, otherwise :ship:

jimabramson commented:
rebased.

@wedaly @frrrances could i get a quick sanity check from either of you on how the new button template is set up / internationalized (toward the bottom of the diff) ?

wedaly commented:
@jimabramson Regarding i18n, are we getting the processor name from the E-Commerce API?  And if so, how is that string being internationalized?

jimabramson commented:
@wedaly for the moment we only support cybersource and always transform that to "Credit Card", which is translated explicitly [here](https://github.com/edx/edx-platform/pull/7777/files#diff-fd4675021d17a3af130c8ded0c74a05cR48).  The working idea for additional processors is to create an explicit mapping of (configuration name) -> (display name) in edx-platform code, so it can be translated in the same way.

frrrances commented on lms/static/sass/_developer.scss:
```diff
@@ -50,3 +50,22 @@
+
+    p, span {
+      line-height: 46px;
+      font-weight: bolder;
```
Use the font-weight mixin here: 

```
@extend %t-ultrastrong;
```

wedaly commented:
@jimabramson That makes sense, thanks for the explanation!  :+1:

frrrances commented on lms/static/sass/_developer.scss:
```diff
@@ -50,3 +50,22 @@
+    float: left;
+
+    p, span {
+      line-height: 46px;
```
remove line height and add padding to parent.

frrrances commented on lms/static/sass/_developer.scss:
```diff
@@ -50,3 +50,22 @@
+div.payment-buttons {
+
+  div.purchase {
+    float: left;
```
add padding: 
```
padding: ($baseline*.75) 0;
```

frrrances commented on lms/static/sass/_developer.scss:
```diff
@@ -50,3 +50,22 @@
+    p, span {
+      line-height: 46px;
+      font-weight: bolder;
+      color: rgb(60, 60, 60);  // fixes the word "price"
```
use color variable: 
```
color: $base-font-color;
```

frrrances commented on lms/static/sass/_developer.scss:
```diff
@@ -50,3 +50,22 @@
 }
+
+// for verify_student/make_payment_step.underscore
+div.payment-buttons {
```
please remove all html elements (div, p, a, span) from css - just use classnames.

frrrances commented on lms/static/sass/_developer.scss:
```diff
@@ -50,3 +50,22 @@
+
+  a.payment-button {
+    float: right;
+    margin-left: 10px;
```
Use baseline and RTL mixin: 
```
@include margin-left( ($baseline/2) );
```

frrrances commented:
I left some css notes. HTML looks good! :+1: once you address the css changes.

jimabramson commented:
test this please
